server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/MailDelivrer.java

```diff
@@ -118,14 +123,23 @@ private Domain retrieveTargetHostname(Mail mail) {
         return rcpt.getDomain();
     }
 
-    private ExecutionResult doDeliver(Mail mail, InternetAddress[] addr, Iterator<HostAddress> targetServers) throws MessagingException {
+    private ExecutionResult doDeliver(Mail mail, ImmutableSet<InternetAddress> addr, Iterator<HostAddress> targetServers) throws MessagingException {
         MessagingException lastError = null;
 
+        ImmutableSet<InternetAddress> targetAddresses = ImmutableSet.copyOf(addr);
+        ImmutableSet.Builder<InternetAddress> deliverySuccess = ImmutableSet.builder();
+
         while (targetServers.hasNext()) {
             try {
-                return mailDelivrerToHost.tryDeliveryToHost(mail, addr, targetServers.next());
+                Collection<InternetAddress> remainingAddresses = Sets.difference(targetAddresses, deliverySuccess.build());
+                return mailDelivrerToHost.tryDeliveryToHost(mail, remainingAddresses, targetServers.next());
             } catch (SendFailedException sfe) {
                 lastError = handleSendFailExceptionOnMxIteration(mail, sfe);
+
+                ImmutableSet<InternetAddress> es = StreamUtils.ofNullable(sfe.getValidSentAddresses())
```

Comment:
can you find a good name instead of `es` ?

Reply:
For sure!